### PR TITLE
sis-scraper: optimize quacs-rs bitvecs by pruning unnecessary conflic…

### PR DESCRIPTION
**Pull request by johnnyapol**
…t blocks

This change introduces an optimization that allows quacs-rs to require fewer
uint64_t 'bitvecs' by not considering slots that have either 0 or exactly 1 classes in them.
The reasoning for this is there can never be a conflict in that case, so we're just wasting memory keeping them around.

Testing on Spring 2021 data, this reduces the number of u64s from 10 to 7, a 30% reduction

### CODE REVIEW

File: sis_scraper/main.py [hunk 1]
```diff
@@ -427,7 +427,21 @@ def optimize_column_ordering(data, num_columns=3):
                                         ] = 1
 
                     conflicts[section["crn"]] = "".join(str(e) for e in conflict)
-
+        # If 0 courses, or exactly one course, occupies a slot in the conflict bitstring, we can safely remove it to save space
+        # as that implies no conflicts are possible in that block.
+        # The following code computes a list of candidates that fit this criteria
+        prune = [bit_index for bit_index in range(0, BIT_VEC_SIZE)
+                if sum(int(conflicts[section_crn][bit_index]) for section_crn in conflicts) in [0, 1]]
```
Comment (benjaminrsherman) [inline sis_scraper/main.py]: The condition here would be a little easier to read as `<= 1` instead of `in [0, 1]`.

File: sis_scraper/main.py [hunk 2]
```diff
@@ -427,7 +427,21 @@ def optimize_column_ordering(data, num_columns=3):
                                         ] = 1
 
                     conflicts[section["crn"]] = "".join(str(e) for e in conflict)
-
+        # If 0 courses, or exactly one course, occupies a slot in the conflict bitstring, we can safely remove it to save space
+        # as that implies no conflicts are possible in that block.
+        # The following code computes a list of candidates that fit this criteria
+        prune = [bit_index for bit_index in range(0, BIT_VEC_SIZE)
```
Comment (benjaminrsherman) [inline sis_scraper/main.py]: I'd like it if this variable name was changed to something a little more descriptive... "unnecessary_indices", maybe?

**Review by benjaminrsherman — CHANGES_REQUESTED** [review]
Overall this is a great change, but there are a few nitpicks in some places.

File: sis_scraper/main.py (hunk 1 above)
Comment (johnnyapol) [reply sis_scraper/main.py]: Applied in latest change (force push amend)

File: sis_scraper/main.py (hunk 2 above)
Comment (johnnyapol) [reply sis_scraper/main.py]: Applied in latest change (force push amend)

**Review by benjaminrsherman — APPROVED** [review]

**Pull request merged**